### PR TITLE
Validations

### PR DESCRIPTION
**Pull request by Danieldotcomcoder**
In this PR I did the following points :

- Added Validations for the model methods
- Added unit specs for all of your models' methods and validations.
- Added flash messages in the create actions in all your controllers.
- Made sure that the N+1 problem is solved when fetching all posts and their comments for a user.

### CODE REVIEW

**Review by Lameck1 — CHANGES_REQUESTED** [review]
Hi, @Danieldotcomcoder, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

# Highlights
- Professional README ✔️
- Clean code ✔️ 
- Good commit messages ✔️ 

## Required Changes ♻️
#### These comments are left here on the review summary since you did not work on these sections in your code.

- [ ] Please ensure that the `likes#create` action also uses flash messages for a successful and not successful path. 
- [ ] It would be nice if we care of N+1 problems in the posts_controller while fetching all posts to all fetch all their related comments. This is as required of us as quoted below.

  > - Make sure that the N+1 problem is solved when fetching all posts and their comments for a user. 
  
  Example:
  
  ```ruby
  @posts = @user.posts.includes(:comments)
  ```
  
  One last thing, in the index action of the `posts_controller` you are only returning the recent posts but we do need to display all user's posts. Consider refactoring this too.

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**Please, remember to tag me in your question so I can receive the notification.**

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/controllers/posts_controller.rb [hunk 1]
```diff
@@ -23,9 +23,11 @@ def create
     respond_to do |format|
       format.html do
         if new_p.save
+          flash[:success] = 'Post saved successfully'
           redirect_to "/users/#{current_user.id}/posts/", notice: 'post created!'
         else
           render :new
+          flash.now[:error] = 'Error: Post could not be saved'
```
Comment (DeliceLydia) [inline app/controllers/posts_controller.rb]: - [ ] `posts#create`, `comments#create`, and `likes#create` actions should use flash messages for a successful and not successful path you can check this [link](https://www.rubyguides.com/2019/11/rails-flash-messages/) to figure out what is missing

**Review by DeliceLydia — CHANGES_REQUESTED** [review]
## Hi @Danieldotcomcoder, 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.


 ## Required Changes ♻️
_Check the comments under the review._



### Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/controllers/posts_controller.rb (hunk 1 above)
Comment (Danieldotcomcoder) [reply app/controllers/posts_controller.rb]: Thank you , I didn't actually understand what you mean by this

**Review by Lameck1 — APPROVED** [review]
Hi, @Danieldotcomcoder,

Tasks for this project have been successfully implemented 💯.

# Highlights
- Professional README ✔️
- Clean code ✔️ 
- Good commit messages ✔️ 

All the best with the next project.

You are free to merge this branch :shipit:
Congratulations! 🎉 

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**Please, remember to tag me in your question so I can receive the notification.**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**